src/main/scala/battleship/GameEngine.scala

```diff
@@ -3,19 +3,24 @@ package battleship
 import battleship.entity._
 
 /**
-  * TODO: Add comment
+  * Game engine which supposed to support shooting and other game logic
   */
 object GameEngine {
 
   def shoot(attackingPlayer: Player, defendingPlayer: Player, hitTarget: Coordinate): Round = {
-    val maybeDecker: Option[Decker] = defendingPlayer.map.ships.flatMap(ship => ship.deckers).find(decker => decker.position == hitTarget)
+    val maybeShip = defendingPlayer.map.ships.find(ship => ship.deckers.exists(decker => decker.position == hitTarget))
 
-    if (maybeDecker.isDefined) {
-//      maybeDecker.get.alive = false
-      Round(attackingPlayer, defendingPlayer, hitTarget, successful = true)
-    } else {
-      Round(attackingPlayer, defendingPlayer, hitTarget, successful = false)
-    }
+    val hitDefendingPlayer = maybeShip.map(ship =>
+      defendingPlayer.copy(map =
+        defendingPlayer.map.copy(ships =
+          defendingPlayer.map.ships.map {
+            case `ship` => ship.copy(deckers = ship.deckers.map {
+              d => if (d.position == hitTarget) d.copy(alive = false) else d
+            });
+            case x => x
```

Comment:
looks like a redundant expression.
it also makes sense to split this nested logic to make it more clear and testable.

```
    val hitDefendingPlayer = maybeShip.map(ship =>
      defendingPlayer.copy(map =
        defendingPlayer.map.copy(ships =
          defendingPlayer.map.ships
            .map(setDeckers(_, hitTarget)))))

    def setDeckers(ship: Ship, hitTarget: Coordinate): Ship = {
      ship.copy(deckers = ship.deckers.map {
        d => if (d.position == hitTarget) d.copy(alive = false) else d
      })
    }
```

Reply:
To make such transformations much easier: http://julien-truffaut.github.io/Monocle/
This is an implementation of the haskell's lenses.

Reply:
Looks simpler, thanks
here eaca855

Reply:
@vitold 
Used monocle for this case, please take a look: 0d8d2c9

Reply:
looks good, but could be even better:

to reduce boilerplate you could use @Lenses("_") annotations to generate lenses
without manual lenses and traversal generation you just need next lines:

```
(Player._map composeLens
      BattleMap._ships composeTraversal each composeLens
      Ship._deckers composeTraversal each) modify(tryHitDecker(_, hitTarget))
```
and don't forget to add compiler pluging for macro paradise:
```
addCompilerPlugin("org.scalamacros" %% "paradise" % "2.1.0" cross CrossVersion.full)
```

Reply:
Done here 6c9ca6e